### PR TITLE
[dynamo] Add LazyVariableTracker

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112961
* #112962
* #111726
* #111725
* #111415
* __->__ #111306

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Comment by Chillee** [comment]
Is this for improving trace times?

**Comment by jansel** [comment]
The idea would be to make all guards eagerly installed on `tx.ouput`, and remove the per-VariableTracker `.guards`.  There have been a bunch of bugs related to that.

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -436,9 +437,10 @@ def index_source(key):
                     return key
 
             result = {
-                k: VariableBuilder(
-                    self.tx, GetItemSource(self.get_source(), index_source(k))
-                )(value[k]).add_guards(guards)
+                k: LazyVariableTracker(
```
Comment (voznesenskym) [inline torch/_dynamo/variables/builder.py:440]: Feels like export should always just realize here too?

**Comment by voznesenskym** [comment]
Fixed model failures, added comments, todos, minor stuff. Looking at eager guards now.

File: torch/_dynamo/variables/lazy.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        if self.mutable_local is None:
+            self.mutable_local = LazyMutableLocal()
+        # TODO(voz) - Support SourcelessBuilder during realize
+        assert (
+            self.source
+        ), "Illegal construction. LazyVariableTracker deferred creation utilizes VariableBuilder."
+
+    def realize(self) -> VariableTracker:
+        """Force construction of the real VariableTracker"""
+        if self.mutable_local.vt is None:
```
Comment (voznesenskym) [inline torch/_dynamo/variables/lazy.py:43]: `self.mutable_local.vt` feels ripe for a @property ;)

**Review by voznesenskym — APPROVED** [review]
image

File: torch/_dynamo/variables/builder.py (hunk 1 above)
Comment (jansel) [reply torch/_dynamo/variables/builder.py:440]: I think the one in symbolic_convert.py should cover it.

File: torch/_dynamo/variables/lazy.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    """
+
+    _nonvar_fields = ["_value"]
+
+    def __init__(self, _value, source, **kwargs):
+        super().__init__(source=source, **kwargs)
+        self._value = _value
+        if self.mutable_local is None:
+            self.mutable_local = LazyMutableLocal()
+        # TODO(voz) - Support SourcelessBuilder during realize
```
Comment (jansel) [inline torch/_dynamo/variables/lazy.py:36]: No point in doing this, since SourcelessBuilder can't create guards

Comment (voznesenskym) [reply torch/_dynamo/variables/lazy.py:36]: Fair point. I was thinking there might be value for LazyVariableTracker for intermediary types, but, thinking more, its not that useful.

File: torch/_dynamo/variables/tensor.py [hunk 4]
```diff
@@ -121,19 +121,6 @@ def as_proxy(self):
...
-                return issubclass(self.python_type(), ty)
-
-            dtypes = tensortype_to_dtype[ty]
-            return self.dtype in dtypes
-
-        if type(tensor_type) is tuple:
-            return any(check_type(ty) for ty in tensor_type)
-        else:
-            return check_type(tensor_type)
-
```
Comment (jansel) [inline torch/_dynamo/variables/tensor.py:136]: @voz did you remove this?  If so why?

Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py:136]: It's moved into `torch/_dynamo/variables/builtin.py` - this is the only place that has this pattern, and it was causing a bug, where we were trying to invoke this incorrectly.

File: torch/_dynamo/variables/builder.py [hunk 5]
```diff
@@ -443,9 +444,10 @@ def index_source(key):
                     return key
 
             result = {
-                k: VariableBuilder(
-                    self.tx, GetItemSource(self.get_source(), index_source(k))
-                )(value[k]).add_guards(guards)
+                k: LazyVariableTracker.create(
+                    value[k],
+                    source=GetItemSource(self.get_source(), index_source(k)),
+                ).add_guards(guards)
```
Comment (lezcano) [inline torch/_dynamo/variables/builder.py:450]: How come this is used in particular for values of dictionaries?

Comment (jansel) [reply torch/_dynamo/variables/builder.py:450]: The next PR in the stack turns it on in more places.  Dict keys seem like something more likely to be unused.

**Comment by zhxchen17** [comment]
@zhxchen17 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50979121).

**Pull request closed without merging**